internal/controller/postgrescluster/snapshots.go

```diff
@@ -133,49 +148,295 @@ func (r *Reconciler) reconcileVolumeSnapshots(ctx context.Context,
+		return pvc, errors.WithStack(err)
+	}
+
+	// If the restore job failed, create a warning event.
+	if restoreJob.Status.Failed == 1 {
+		r.Recorder.Event(cluster, corev1.EventTypeWarning,
+			"DedicatedSnapshotVolumeRestoreJobError", "restore job failed, check the logs")
+		return pvc, nil
+	}
+
+	// If we made it here, the restore job is still running and we should do nothing.
+	return pvc, err
+}
+
+// createDedicatedSnapshotVolume creates/updates/gets the dedicated snapshot volume.
+// It expects that the volume name and GVK has already been set on the pvc that is passed in.
+// TODO(dsessler7): The above fact makes this method feel a little awkward to me.
+// Not sure this code really needs its own method. I'm tempted to just move it back
+// to reconcileDedicatedSnapshotVolume.
+func (r *Reconciler) createDedicatedSnapshotVolume(ctx context.Context,
+	cluster *v1beta1.PostgresCluster, labelMap map[string]string,
+	pvc *corev1.PersistentVolumeClaim,
+) (*corev1.PersistentVolumeClaim, error) {
+	var err error
+
+	// An InstanceSet must be chosen to scale resources for the dedicated snapshot volume.
+	// TODO: We've chosen the first InstanceSet for the time being, but might want to consider
+	// making the choice configurable.
+	instanceSpec := cluster.Spec.InstanceSets[0]
+
+	pvc.Annotations = naming.Merge(
+		cluster.Spec.Metadata.GetAnnotationsOrNil(),
+		instanceSpec.Metadata.GetAnnotationsOrNil())
+
+	pvc.Labels = naming.Merge(
+		cluster.Spec.Metadata.GetLabelsOrNil(),
+		instanceSpec.Metadata.GetLabelsOrNil(),
+		labelMap,
+	)
+
+	err = errors.WithStack(r.setControllerReference(cluster, pvc))
+	if err != nil {
+		return pvc, err
+	}
+
+	pvc.Spec = instanceSpec.DataVolumeClaimSpec
+
+	// Set the snapshot volume to the same size as the pgdata volume. The size should scale with auto-grow.
+	r.setVolumeSize(ctx, cluster, pvc, instanceSpec.Name)
+
+	// Clear any set limit before applying PVC. This is needed to allow the limit
+	// value to change later.
+	pvc.Spec.Resources.Limits = nil
+
+	err = r.handlePersistentVolumeClaimError(cluster,
+		errors.WithStack(r.apply(ctx, pvc)))
+	if err != nil {
+		return pvc, err
+	}
+
+	return pvc, err
+}
+
+// dedicatedSnapshotVolumeRestore creates a Job that performs a restore into the dedicated
+// snapshot volume.
+// This function is very similar to reconcileRestoreJob, but specifically tailored to the
+// dedicated snapshot volume.
+func (r *Reconciler) dedicatedSnapshotVolumeRestore(ctx context.Context,
+	cluster *v1beta1.PostgresCluster, dedicatedSnapshotVolume *corev1.PersistentVolumeClaim,
+	instances *observedInstances, clusterVolumes []corev1.PersistentVolumeClaim,
+	backupJob *batchv1.Job,
+) error {
+
+	pgdata := postgres.DataDirectory(cluster)
+	repoName := backupJob.GetLabels()[naming.LabelPGBackRestRepo]
+
+	opts := []string{
+		"--stanza=" + pgbackrest.DefaultStanzaName,
+		"--pg1-path=" + pgdata,
+		"--repo=" + regexRepoIndex.FindString(repoName),
+		"--delta",
+		// "--link-all",
```

Comment:
🤔 IIUC, we don't want a symlink to a WAL volume that doesn't exist. Perhaps instead of saying nothing about `--link-all` we should say "no" symlinks, however that's spelled. `x=false` ? `--no-x` ?

❓ What happens with tablespaces?

Reply:
Thanks for bringing this up. Brian was using `--link-all` in his script. I wasn't sure if we wanted to use it which is why it is there, but commented out.

With tablespaces, we do the same thing that we do in our other `RestoreCommand` function...